pf/tests/testdata/updateprogram/Pulumi.yaml

```diff
@@ -38,7 +38,7 @@ outputs:
   testOptionalString__expect: |-
     input2
   testSetOptionals__actual: ${set.setOptionals}
-  testSetOptionals__expect: [ "${r.requiredInputStringCopy}", a, b ]
+  testSetOptionals__expect: [ a, b, "${r.requiredInputStringCopy}" ]
```

Comment:
This change caused set elements to be reordered in the list projection to Pulumi. While this should theoretically be alright, as the subsequent Diff recognizes that sets are sets and should not be sensitive to element order, perhaps there are some unintended consequences?

Should the bridge define some guarantees on the canonical set ordering when projecting elements to Pulumi array property values?

Reply:
I think the bridge needs to define a constant ordering since we expose this data as a list. We shouldn't need to guarantee anything about the ordering, except consistency.

Its not unreasonable to assume that `new Resource("one", opt=set[0]); new Resource("two", opt=set[1])` is safe between updates, where a re-order would break this.

Reply:
I'm not sure I grok the example, could you expand. 

Since bridged providers customize Diff method sending re-ordered versions of the set should result in a no-changes diff. But I have a hunch somewhere we run into trouble anyway. Would be good to file an issue figuring out where. 

Perhaps with ignoreChanges paths is one place. 

FWIW this PR seems to move us from "user-defined" ordering to "TF-defined" consistent ordering, indirectly, based on the example changes. There's some subtleties with this. It probably is affected by TF hashing functions and when they change upstream this is likely to reorder sets.

There are also some serious issues with sets on the SDKV2 based providers. I was recently looking at this one:

https://github.com/pulumi/pulumi-terraform-bridge/pull/1255

Reply:
https://github.com/pulumi/pulumi-terraform-bridge/pull/1276#pullrequestreview-1518268522 has a better example. Basically, external users can rely on the ordering of pulumi lists (and tf sets) such that re-ordering will always cause problems, no matter what the bridge does.

Reply:
If we believe that TF-defined sets are better (more consistent, easier to maintain) then I'm ok moving to them as long as we (1) recognize it is a breaking change and (2) put tests and documentation locking that ordering in.